### PR TITLE
Remaster dice game

### PR DESCRIPTION
**Pull request by cyphersnake**
- Make math part great again
- Move all constants in separated namespace
- Refactor and code improvements

### CODE REVIEW

File: contracts/src/dice.cpp [hunk 1]
```diff
@@ -2,73 +2,76 @@
 
 namespace dice {
 
-void dice::check_params(uint64_t ses_id) {
-    eosio::check(get_param_value(ses_id, min_bet_param_type) != std::nullopt, "absent min bet param");
-    eosio::check(get_param_value(ses_id, max_bet_param_type) != std::nullopt, "absent max bet param");
-    eosio::check(get_param_value(ses_id, max_payout_param_type) != std::nullopt, "absent max payout param");
+void dice::check_params(uint64_t ses_id) const {
```
Comment (algys) [inline contracts/src/dice.cpp:5]: also need to add params relations checks (max_bet > min_bet, max_payout > max_bet, etc..)

File: contracts/src/dice.cpp [hunk 2]
```diff
@@ -2,73 +2,76 @@
...
-    rolls.emplace(get_self(), [&](auto& row){
+    const auto& number = params[0];
+    rolls.emplace(get_self(), [&](auto& row) {
         row.ses_id = ses_id;
-        row.number = params[0];
+        row.number = number;
     });
 
-    update_max_win(ses_id, calc_max_win(ses_id, params[0]));
+    update_max_win(ses_id, asset(get_win_coefficient(number), core_symbol));
```
Comment (algys) [inline contracts/src/dice.cpp]: mb here should be `get_win_payout` instead of `get_win_coefficient` ?

Comment (algys) [inline contracts/src/dice.cpp]: + add tests about `max_win`

**Review by algys — CHANGES_REQUESTED** [review]

File: contracts/src/dice.cpp (hunk 2 above)
Comment (cyphersnake) [reply contracts/src/dice.cpp]: I ask you about it, you tell me use relative coefficient. I can change it to absolute value.

Comment (cyphersnake) [reply contracts/src/dice.cpp]: At now - all tests about `max_win`, I just change expected values actually.

File: contracts/src/dice.cpp [hunk 3]
```diff
@@ -2,73 +2,86 @@
...
 
-    eosio::print("rand num: ", rand_number, "\n");
+    auto payout = zero_asset;
 
-    if (roll.number >= rand_number) { //loose
-        finish_game(ses_id, zero_asset);
-        return;
-    }
+    const auto bet_number = rolls.get(ses_id).number;
+    if (bet_number < actual_number) { // win
```
Comment (Eenae) [inline contracts/src/dice.cpp]: `<=`

**Review by algys — APPROVED** [review]

**Comment by Eenae** [comment]
Ok, now the average player profit is `- house_edge * deposit`

**Review by Eenae — APPROVED** [review]

**Pull request merged**